kahypar/partition/bin_packing/bin_packer.h

```diff
@@ -0,0 +1,166 @@
+    return previous;
+  }
+
+class IBinPacker {
+ public:
+  IBinPacker(const IBinPacker&) = delete;
+  IBinPacker(IBinPacker&&) = delete;
+  IBinPacker& operator= (const IBinPacker&) = delete;
+  IBinPacker& operator= (IBinPacker&&) = delete;
+
+  void prepacking(Hypergraph& hg, Context& context, const BalancingLevel level) {
+    prepackingImpl(hg, context, level);
+  }
+
+  std::vector<PartitionID> twoLevelPacking(const Hypergraph& hg, const Context& context, const std::vector<HypernodeID>& nodes,
+                                           const HypernodeWeight max_bin_weight) {
+    return twoLevelPackingImpl(hg, context, nodes, max_bin_weight);
+  }
+
+  virtual ~IBinPacker() = default;
+
+ protected:
+  IBinPacker() = default;
+
+ private:
+  virtual void prepackingImpl(Hypergraph& hg, Context& context, const BalancingLevel level) = 0;
+  virtual std::vector<PartitionID> twoLevelPackingImpl(const Hypergraph& hg, const Context& context, const std::vector<HypernodeID>& nodes,
+                                                       const HypernodeWeight max_bin_weight) = 0;
+};
+
+template< class BPAlgorithm = WorstFit >
+class BinPacker final : public IBinPacker {
+ public:
+  BinPacker() = default;
+
+ private:
+  void prepackingImpl(Hypergraph& hg, Context& context, const BalancingLevel level) override {
+    ASSERT(!hg.containsFixedVertices(), "No fixed vertices allowed before prepacking.");
+    ASSERT(level != BalancingLevel::STOP, "Invalid balancing level: STOP");
+
+    PartitionID rb_range_k = context.partition.rb_upper_k - context.partition.rb_lower_k + 1;
+    HypernodeWeight max_bin_weight = floor(context.initial_partitioning.current_max_bin * (1.0 + context.initial_partitioning.bin_epsilon));
+
+    if (level == BalancingLevel::optimistic) {
+      bin_packing::calculateHeuristicPrepacking<BPAlgorithm>(hg, context, rb_range_k, max_bin_weight);
+    } else if (level == BalancingLevel::guaranteed) {
+      for (size_t i = 0; i < static_cast<size_t>(context.initial_partitioning.k); ++i) {
+        HypernodeWeight lower = context.initial_partitioning.perfect_balance_partition_weight[i];
+        HypernodeWeight& border = context.initial_partitioning.upper_allowed_partition_weight[i];
+        HypernodeWeight upper = context.initial_partitioning.num_bins_per_part[i] * max_bin_weight;
+
+        // TODO(maas) Here, some magic numbers are used to check that the different borders for the part weights work
+        // for the prepacking algorithm. Unfortunately, I do not know of a more general or elegant solution.
+        // The basic problem is, if <border> (the allowed partition weight) is too close to <upper>, the prepacking
+        // algorithm performs badly in the sense that almost every vertex will be prepacked, thus resulting in a
+        // bisection which is indeed balanced, but probably has a really poor cut.
+        if (upper - border < (border - lower) / 10) {
```

Comment:
Unfortunately no, as I had no good idea how a good experimental setup for these numbers could look like and there were many other things that needed to be finished for my thesis. The numbers are just some values that seemed sensible based on my theoretical understanding of the algorithm.

The problem is that the prepacking algorithm is usually called for no more than 1 or 2 bisections on the real-world instances I used, possibly not invoking this case at all. Therefore, I suspect a really fine tuned setup is required to measure any notable effect on result quality (perhaps it might help a bit to turn k-way refinement off and only evaluate the initial partitioning result).